### PR TITLE
UDAF: Extend more args to `state_fields` and `groups_accumulator_supported` and introduce `ReversedUDAF`

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #10391.

## Rationale for this change

This PR is pulled from #10484 and it has a similar rationale in #10391 but the changes in this PR are for #10484

### CODE REVIEW

File: datafusion/expr/src/function.rs [hunk 1]
```diff
@@ -66,22 +69,39 @@ pub struct AccumulatorArgs<'a> {
...
-        sort_exprs: &'a [Expr],
-    ) -> Self {
-        Self {
-            data_type,
-            schema,
-            ignore_nulls,
-            sort_exprs,
-        }
-    }
+pub struct StateFieldsArgs<'a> {
```
Comment (alamb) [inline datafusion/expr/src/function.rs:90]: Can we please add doc comments to this explaining what it is for?

File: datafusion/expr/src/function.rs [hunk 2]
```diff
@@ -66,22 +69,39 @@ pub struct AccumulatorArgs<'a> {
...
+    /// Whether the aggregate function is distinct.
+    pub is_distinct: bool,
+
+    /// The input type of the aggregate function.
+    pub input_type: &'a DataType,
+}
+
+/// [`GroupsAccumulatorSupportedArgs`] contains information to determine if an
+/// aggregate function supports the groups accumulator.
+pub struct GroupsAccumulatorSupportedArgs {
```
Comment (alamb) [inline datafusion/expr/src/function.rs:82]: I know the current implementation of `groups_accumulator_supported` only needs these two fields, but I can see in the future needing more of them

Is there any reason we can't just pass `AccumulatorArgs` to `groups_accumulator_supported`?

**Review by alamb — COMMENTED** [review]
Thanks @jayzhan211  -- I think this is a very nice PR 🙏 

The only thing I think we should consider is reusing AccumulatorArgs rather than adding `GroupsAccumulatorSupportedArgs`

File: datafusion/expr/src/function.rs (hunk 2 above)
Comment (jayzhan211) [reply datafusion/expr/src/function.rs:82]: We could do this.

**Review by alamb — APPROVED** [review]
Thanks @jayzhan211

**Pull request merged**